### PR TITLE
Added "--track_promote_to" option for promoting APK between tracks.  …

### PR DESCRIPTION
**Pull request by sschendel**
Finally got around to adding track promotion feature to new mono repo.

See issue #4563. 

See https://github.com/fastlane-old/supply/pull/52.

### CODE REVIEW

File: supply/lib/supply/options.rb [hunk 1]
```diff
@@ -111,7 +111,16 @@ def self.available_options
...
                                      is_string: false,
-                                     default_value: false)
+                                     default_value: false),
+        FastlaneCore::ConfigItem.new(key: :track_promote_to,
+                                     env_name: "SUPPLY_TRACK_PROMOTE_TO",
+                                     optional: true,
+                                     description: "The Track to promote to: production, beta, alpha or rollout",
+                                     verify_block: proc do |value|
+                                       available = %w(production beta alpha rollout)
+                                       raise "Invalid value '#{value}', must be #{available.join(', ')}".red unless available.include? value
```
Comment (asfalcone) [inline supply/lib/supply/options.rb]: This should be the new style of error handling `UI.user_error! ...` instead of `raise ... red`

File: supply/lib/supply/options.rb [hunk 2]
```diff
@@ -111,7 +111,16 @@ def self.available_options
...
                                      description: "Whether to skip uploading SCREENSHOTS",
                                      is_string: false,
-                                     default_value: false)
+                                     default_value: false),
+        FastlaneCore::ConfigItem.new(key: :track_promote_to,
+                                     env_name: "SUPPLY_TRACK_PROMOTE_TO",
+                                     optional: true,
+                                     description: "The Track to promote to: production, beta, alpha or rollout",
+                                     verify_block: proc do |value|
+                                       available = %w(production beta alpha rollout)
```
Comment (asfalcone) [inline supply/lib/supply/options.rb]: Can you extract this as a variable this array of valid tracks is also used [here](https://github.com/sschendel/fastlane/blob/promote_track/supply/lib/supply/options.rb#L19)

File: supply/lib/supply/uploader.rb [hunk 3]
```diff
@@ -25,11 +25,21 @@ def perform_upload
...
       upload_binaries unless Supply.config[:skip_upload_apk]
 
+      promote_track if Supply.config[:track_promote_to]
+
       UI.message("Uploading all changes to Google Play...")
       client.commit_current_edit!
       UI.success("Successfully finished the upload to Google Play")
     end
 
+    def promote_track
```
Comment (asfalcone) [inline supply/lib/supply/uploader.rb]: Is it possible to reuse the code [here](https://github.com/sschendel/fastlane/blob/promote_track/supply/lib/supply/uploader.rb#L143)?

Can you summarize how the promote to track option is different than calling `supply` with a different track than is currently set in google play? I just want to make sure that this feature is neccessary and not something that can already be done with the existing set up 👍

**Comment by asfalcone** [comment]
@sschendel Thanks for the PR! I've added some comments on things that need to be cleaned up before this could be merged.

Please also be sure that you've responsed to the feedback from @lacostej on the [original PR](https://github.com/fastlane-old/supply/pull/52)

I also left a request for commentary about how/when this feature would be used.  I want to make sure this feature isn't confusing for `supply` users 👍

File: supply/lib/supply/uploader.rb (hunk 3 above)
Comment (sschendel) [inline supply/lib/supply/uploader.rb]: The Play API does not allow upload of an APK version code if it already exists as active version on another track.    

The use case we have (and I think it may be typical) is to upload APK to a test track (we use beta), test it, and finally promote that version to production.  This cannot be done with existing tool (as far as I can tell).  Even if it were possible to upload same version to another track - it is convenient to NOT have to upload the same APK a second time.  In our situation (with 60+ apps) that can take much longer than just commanding the API to promote the APK between tracks.

The current supply framework really centers on the act of "performing an upload".  Commanding the Play API to promote already uploaded assets between tracks doesn't fit perfectly.  I tried to fit it best I could while solving the real problem I had.  Consider it a functional feature that is open to improvement to anyone with ideas.

Summarized feature:
"--track_promote_to=" works with the "--track=" parameter to command the Play API to promote existing Play track APK version(s) (--track param value) to a new track (--track_promote_to value).

Comment (asfalcone) [inline supply/lib/supply/uploader.rb]: Thanks, great summary! Can you add a small section to the [README](https://github.com/fastlane/fastlane/blob/master/supply/README.md) after **Changelog** with the summaried feature and a short explain of how it can be used?

**Comment by asfalcone** [comment]
This looks great. I added 1 more request: update the documentation. Then we'll be ready to merge 🚀

**Comment by sschendel** [comment]
Added the README section you asked for, @asfalcone.

**Comment by asfalcone** [comment]
Thanks @sschendel 💯  So glad we finally got this in!!!
